### ISSUE TITLE
gl122 test failure

### ISSUE THREAD

**Issue by joebo** [jsoftware/j-playground#34]
This test case fails https://github.com/jsoftware/j-playground/blob/j903_wasm/test/g122.ijs#L126

```
   match =: 0:`(*./@:=&,)@.(-:&$)
   test  =: -: = match
   (}.a,5)   test a=:?100$2
```
[Playground](https://jsoftware.github.io/j-playground/bin/html2/#code=match%20%3D%3A%200%3A%60(*.%2F%40%3A%3D%26%2C)%40.(-%3A%26%24)%0Atest%20%20%3D%3A%20-%3A%20%3D%20match%0A(%7D.a%2C5)%20%20%20test%20a%3D%3A%3F100%242)

But passes in j9   

I don't understand the code of match well enough without a fair amount of research to track it down. What does match do?

**Comment by jonghough** [comment]
The test seems to be
_if two arrays are not identical, then there must be at least one different item in the arrays._

If we say
```
 b=:}. a,5
 ```
 
then in Jplayground

```
match =: 0:`(*./@:=&,)@.(-:&$)
b match a
1
```

This is the problem. a and b now have the same shape but definitely one item not equal (the last item).
In `J901` and `J805` match works correctly.

If we add superfluous braces in Jplayground
```
match =: 0:`(*./@:(=&,))@.(-:&$)
```
the test works.

It seems the parsing rules are broken in some specific way.

**Comment by joebo** [comment]
Thanks jon -- your triage of the issue is very helpful. Glad to hear it passes with superflous parentheses but odd indeed.

**Comment by jonghough** [comment]
Actually I think I spoke too soon.

Try
``` 
match2 =: (*./@:=)
a=:?4$2
b=:}. a,5
a
0 0 0 1
b
0 0 1 5
b match2 a
1  NB. this is wrong!
```

It is wrong in Jplayground, but fine in J901, J805.

Even in Jplayground the verb "looks" like it should behave correctly/

```
 9!:3 (2)
 match2
 
 +------+--+-+
|+--+-+|@:|=|
||*.|/||  | |
|+--+-+|  | |
+------+--+-+
```

Perhaps `*./` has an issue. But maybe not, as `(+/@:=)` also gives wrong results in JPlayground.

**Comment by jonghough** [comment]
Actually I think I spoke too soon.

Try
```
match2 =: (*./@:=)
a=:?4$2
b=:}. a,5
a
0 0 0 1
b
0 0 1 5
b match2 a
1  NB. this is wrong!
```

It is wrong in Jplayground, but fine in J901, J805.

Even in Jplayground the verb "looks" like it should behave correctly/
```
 9!:3 (2)
 match2
 
 +------+--+-+
|+--+-+|@:|=|
||*.|/||  | |
|+--+-+|  | |
+------+--+-+
```
Perhaps `*./` has an issue. But maybe not, as `(+/@:=)` also gives wrong results in JPlayground.

Weirder still,

```
match3=: (+/@:=)
a match3 b
3
b match3 a
0
```

i.e. switching x and y changes the result.

**Comment by rdm** [comment]
[Playground](https://jsoftware.github.io/j-playground/bin/html2/#code=match%20%3D%3A%200%3A%60(*.%2F%40%3A%3D%26%2C)%40.(-%3A%26%24)%0Atest%20%20%3D%3A%20-%3A%20%3D%20match%0A(%7D.a%2C5)%20%20%20test%20a%3D%3A%3F100%242) works for me (gives me a zero result), in Microsoft Edge Version 100.0.1185.50 (Official build) (64-bit) and in Vivaldi 5.2.2623.33 (Stable channel) (64-bit)

Has a fix to this bug been deployed?

**Comment by joebo** [comment]
> [Playground](https://jsoftware.github.io/j-playground/bin/html2/#code=match%20%3D%3A%200%3A%60(*.%2F%40%3A%3D%26%2C)%40.(-%3A%26%24)%0Atest%20%20%3D%3A%20-%3A%20%3D%20match%0A(%7D.a%2C5)%20%20%20test%20a%3D%3A%3F100%242) works for me (gives me a zero result), in Microsoft Edge Version 100.0.1185.50 (Official build) (64-bit) and in Vivaldi 5.2.2623.33 (Stable channel) (64-bit)
> 
> Has a fix to this bug been deployed?

A zero result is a failure in the test cases. Success is 1, Failure is 0

**Comment by rdm** [comment]
Oops... for some reason I was thinking I was looking at the result of match -- sorry about that.

I suspect something is wrong with @: here, and:

```
(}.a,5)   (-:;0:`(*./@:]@:=&,)@.(-:&$)) a=:?100$2
+-+-+
|0|0|
+-+-+
```

vs

```
(}.a,5)   (-:;0:`(*./@:=&,)@.(-:&$)) a=:?100$2
+-+-+
|0|1|
+-+-+

```

Specifically, I suspect the changes introduced in (and presumably somewhat fixed since) j903 beta i

https://code.jsoftware.com/wiki/System/ReleaseNotes/J903 

_Complete rewrite of combinations created by hook/fork/@:/&:/&./&.: . Now blocks are freed immediately after the verb they are applied to, provided that verb did not save the value or modify it inplace._

That said, I'm not sure what portability ambiguity would be relevant here, and I would need to set up a good debugging environment to pursue this further.

**Comment by joebo** [comment]
raul, your investigation is very helpful. I also took a stab at making the hacked linux build run the tests. The hacked linux build also fails on this test, so it's either something in the hacks I made to make the code compile under wasm or it's a bug in the code across all platforms. I think the next step here would be to build the same source under the normal build process for linux and see if it's an error.

As of right now, I do not think this is a bug in how J is running under WASM, which is good. Signs point more towards an existing bug or a bug in the my adjustments to the source code or compile time flags.

I will look more at this as time permits

**Comment by joebo** [comment]
This bug is also affecting use of the addon pjson, which has this ```y +./@:e. '"{[' ```, which seems to do nothing on the J Playground

```
NB. this should output 1
'{' +./@:e. '"{[' 
```
[Reproducible J Playground example](https://jsoftware.github.io/j-playground/bin/html2/#code=NB.%20this%20should%20output%201%0A'%7B'%20%2B.%2F%40%3Ae.%20'%22%7B%5B'%20)


from pjson
```
   dec_array
3 : 0
y=. dltb }.}:y
if. 0=#y do. $0 return. end.
if. -. y +./@:e. '"{[' do. ,dec_num y return. end.
dec1 each cutcommas y
)
```

**Comment by rdm** [comment]
Now that is really odd behavior.

It's not an error, but it's not a result either.  Here's an example session where I try to determine what rank the result has (it apparently does not generate a result with a rank), and then try to determine what error was generated but not reported (it does not generate any error). And yet, it's not just a display artifact, because we can see that it does not generate an assignable result:

(I first look at what pre-existing "last error" was present in my session. Then I look for the rank of this result. Then I look to see if there's a new "last error". Then I verify that "last error" is an abstraction which should have relevance here. And, finally, I try seeing if it's a hidden result by assigning the result to a name and inspecting that name.)

```
   13!:11''
3
   13!:12''
|domain error
|   '() '''[:    [:a.
|[-280] emj.ijs
   #$'{' +./@:e. '"{['
   13!:11''
3
   13!:12''
|domain error
|   '() '''[:    [:a.
|[-280] emj.ijs
   1 2+3 4 5
|length error
|   1 2    +3 4 5
|[-0]
   13!:11''
9
   13!:12''
|length error
|   1 2    +3 4 5
|[-0]
   A=:'{' +./@:e. '"{['
   $A
$ A
   nc<'A'
_1

```

**Comment by joebo** [comment]
I agree - really odd. It's something specific to emscripten and not the J source because the following outputs 1 when added to emj.c and built with makefile.linux

```
#if __linux__
 printf("result: %s", em_jdo("'{' +./@:e. '\"{['"));
#endif
```

I also proved that it will show 0 if it doesn't find it eg: ```em_jdo("'A' +./@:e. '\"{['")``` returns 0

**Comment by joebo** [comment]
Resolved with cba98bc31c503b5b2c6836f17312fb19401f191e as a workaround and related to #85 . The workaround disables special code paths found with jtatcompf. The root cause should still be investigated as time permits, but it's more important for the jengine to work correctly on WASM then to be fast with other non-failing special code paths.
